Mathlib/RingTheory/Nilpotent/GeometricallyReduced.lean

```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2025 Dion Leijnse. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dion Leijnse
+-/
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.FieldTheory.IsAlgClosed.AlgebraicClosure
+import Mathlib.LinearAlgebra.TensorProduct.Finiteness
+import Mathlib.RingTheory.Flat.FaithfullyFlat.Basic
+import Mathlib.RingTheory.Henselian
+
+/-!
+# Geometrically reduced algebras
+
+In this file we introduce geometrically reduced algebras.
+For a field `k`, we say that a `k`-algebra `A` is geometrically reduced (`IsGeometricallyReduced`)
+if the tensor product `A ⊗[k] AlgebraicClosure k` is reduced.
+
+## Main results
+- `FlatBaseChangeIsReduced.of_FG`: if `k` is a commutative ring and `A` and `C` are `k`-algebras,
+  such that `C ⊗[k] B` is reduced for all finitely generated `k` subalgebras `B` of `A`, then
+  `C ⊗[k] A` is reduced.
+
+## References
+- See [https://stacks.math.columbia.edu/tag/05DS] for some theory of geometrically reduced algebras.
+  Note that their definition differs from the one here, we still need a proof that these are
+  equivalent.
+
+-/
+
+open TensorProduct
+
+noncomputable section
+variable {k A : Type*} [Field k] [Ring A] [Algebra k A]
+
+/-- The `k`-algebra `A` is geometrically reduced iff its base change to `AlgebraicClosure k` is
+  reduced -/
+@[mk_iff]
+class IsGeometricallyReduced (k A : Type*) [Field k] [Ring A] [Algebra k A]
+    : Prop where
+  reduced_algebraicClosure_tensor : IsReduced ((AlgebraicClosure k) ⊗[k] A)
+
+attribute [instance] IsGeometricallyReduced.reduced_algebraicClosure_tensor
+
+instance (k A K : Type*) [Field k] [Ring A] [Algebra k A] [Field K] [Algebra k K]
+    [IsAlgClosure k K] [IsGeometricallyReduced k A] : IsReduced (K ⊗[k] A) :=
+  isReduced_of_injective
+    (Algebra.TensorProduct.map
+      ((↑(IsAlgClosure.equiv k K (AlgebraicClosure k)) : K →ₐ[k] AlgebraicClosure k)) 1)
+    (Module.Flat.rTensor_preserves_injective_linearMap _
+      <| EquivLike.injective (IsAlgClosure.equiv k K (AlgebraicClosure k)))
+
+lemma isGeometricallyReduced_of_injective {B : Type*} [Ring B] [Algebra k B] (f : A →ₐ[k] B)
+    (hf : Function.Injective f) [IsGeometricallyReduced k B] : IsGeometricallyReduced k A :=
+  ⟨isReduced_of_injective (Algebra.TensorProduct.map 1 f)
+    (Module.Flat.lTensor_preserves_injective_linearMap _ hf)⟩
+
+theorem isReduced_of_isGeometricallyReduced [IsGeometricallyReduced k A] : IsReduced A :=
```

Comment:
```suggestion
variable (k) in
theorem isReduced_of_isGeometricallyReduced [IsGeometricallyReduced k A] : IsReduced A :=
```

Because `k` cannot be deduced from the conclusion `IsReduced A`.